dev/com.ibm.ws.collector.manager/src/com/ibm/ws/collector/manager/buffer/BufferManagerImpl.java

```diff
@@ -0,0 +1,190 @@
+/*******************************************************************************
+ * Copyright (c) 2015, 2017 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.collector.manager.buffer;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.ws.logging.synch.ThreadLocalHandler;
+import com.ibm.wsspi.collector.manager.BufferManager;
+import com.ibm.wsspi.collector.manager.Handler;
+import com.ibm.wsspi.collector.manager.SyncrhonousHandler;
+
+public class BufferManagerImpl extends BufferManager {
+
+	private static final TraceComponent tc = Tr.register(BufferManagerImpl.class);
+
+	private Set<SyncrhonousHandler> synchronizedHandlerSet = new HashSet<SyncrhonousHandler>();
+
+	private final String sourceId;
+	/* Map to keep track of the next event for a handler */
+	private final ConcurrentHashMap<String, HandlerStats> handlerEventMap = new ConcurrentHashMap<String, HandlerStats>();
+
+	public BufferManagerImpl(int capacity, String sourceId) {
+		super(capacity);
+		this.sourceId = sourceId;
+	}
+
+	@Override
+	public void add(Object event) {
+		
+		/*
+		 * Check if we have any synchronized handlers, and write directly to them
+		 */
+		if (!synchronizedHandlerSet.isEmpty()) {
+			for (SyncrhonousHandler synchronizedHandler : synchronizedHandlerSet) {
```

Comment:
48: you may get a ConcurrentModificationException here if handler set is changed after you have created the iterator